src/main/scala/go/variant/Variant.scala

```diff
@@ -0,0 +1,250 @@
+        position = newPosition.some
+      )
+    )
+  }
+
+  // def move(
+  //     situation: Situation,
+  //     from: Pos,
+  //     to: Pos,
+  //     promotion: Option[PromotableRole]
+  // ): Validated[String, Move] = {
+  //   // Find the move in the variant specific list of valid moves
+  //   situation.moves get from flatMap (_.find(m => m.dest == to && m.promotion == promotion)) toValid
+  //     s"Not a valid move: ${from}${to} with prom: ${promotion}. Allowed moves: ${situation.moves}"
+  // }
+
+  def drop(situation: Situation, role: Role, pos: Pos): Validated[String, Drop] =
+    if (dropsVariant)
+      validDrops(situation).filter(d => d.piece.role == role && d.pos == pos).headOption match {
+        case Some(drop) => Validated.valid(drop)
+        case None       => Validated.invalid(s"$situation cannot perform the drop: $role on $pos")
+      }
+    else Validated.invalid(s"$this variant cannot drop $situation $role $pos")
+
+  def pass(situation: Situation): Validated[String, Pass] = Validated.valid(validPass(situation))
+
+  def possibleDrops(situation: Situation): Option[List[Pos]] =
+    if (dropsVariant)
+      validDrops(situation).map(_.pos).some
+    else None
+
+  def possibleDropsByRole(situation: Situation): Option[Map[Role, List[Pos]]] =
+    if (dropsVariant)
+      validDrops(situation)
+        .map(drop => (drop.piece.role, drop.pos))
+        .groupBy(_._1)
+        .map { case (k, v) => (k, v.toList.map(_._2)) }
+        .some
+    else None
+
+  def stalemateIsDraw = false
+
+  def winner(situation: Situation): Option[Player]
+
+  @nowarn def specialEnd(situation: Situation) = false
+
+  @nowarn def specialDraw(situation: Situation) = false
+
+  def materialImbalance(board: Board): Int =
+    board.pieces.values.foldLeft(0) { case (acc, Piece(player, role)) =>
+      Role.valueOf(role).fold(acc) { value =>
+        acc + value * player.fold(1, -1)
+      }
+    }
+
+  // Some variants have an extra effect on the board on a move. For example, in Atomic, some
+  // pieces surrounding a capture explode
+  def hasMoveEffects = false
+
+  def addVariantEffect(drop: Drop): Drop = drop // should we affect score/captures here?
+
+  /** Once a move has been decided upon from the available legal moves, the board is finalized
+    */
+  @nowarn def finalizeBoard(board: Board, uci: format.Uci, captured: Option[Piece]): Board =
+    board
+
+  def valid(board: Board, strict: Boolean): Boolean =
+    Api.validateFEN(Forsyth.exportBoard(board))
+
+  val roles: List[Role] = Role.all
+
+  lazy val rolesByPgn: Map[Char, Role] = roles
+    .map { r =>
+      (r.pgn, r)
+    }
+    .to(Map)
+
+  override def toString = s"Variant($name)"
+
+  override def equals(that: Any): Boolean = this eq that.asInstanceOf[AnyRef]
+
+  override def hashCode: Int = id
+
+  def defaultRole: Role = Role.defaultRole
+
+  def gameFamily: GameFamily
+}
+
+object Variant {
+
+  lazy val all: List[Variant] = List(
```

Comment:
might want to consider reversing this order?

Reply:
sure